python-main.js

```diff
@@ -1075,31 +1075,56 @@ function web_editor(config) {
             $("#command-disconnect > .roundlabel").text(config["translate"]["static-strings"]["buttons"]["command-disconnect"]["label"]);
             $("#command-disconnect").attr("title", config["translate"]["static-strings"]["buttons"]["command-disconnect"]["title"]);
 
-            // Change download to flash
-            $("#command-download").attr("id", "command-flash");
-            $("#command-flash > .roundlabel").text(config["translate"]["static-strings"]["buttons"]["command-flash"]["label"]);
-            $("#command-flash").attr("title", config["translate"]["static-strings"]["buttons"]["command-flash"]["title"]);
-
             if (serial){
-              doSerial();
+                doSerial();
             }
         })
-        .catch(function(err) {
-            console.log("Error connecting: " + err);
-            $("#flashing-overlay-container").css("display", "flex");
-            $("#flashing-info").addClass('hidden');
-
-            // If micro:bit does not support dapjs
-            if (err.message === "No valid interfaces found."){
-                $("#flashing-overlay-error").html('<div>' + err + '</div><div>'+ config["translate"]["webusb"]["err"]["update-req"] +'</div><a href="#" onclick="flashErrorClose()">'+ config["translate"]["webusb"]["close"] +'</a>');
-                return;
-            } else if (err.message === "Unable to claim interface.") {
-                $("#flashing-overlay-error").html('<div>'+ config["translate"]["webusb"]["err"]["clear-connect"] +'</div><a href="#" onclick="flashErrorClose()">'+ config["translate"]["webusb"]["close"] +'</a>');
-                return;
-            }
-
-            $("#flashing-overlay-error").html('<div>' + err + '</div><div>'+ config["translate"]["webusb"]["err"]["restart-microbit"] +'</div><a href="#" onclick="flashErrorClose()">'+ config["translate"]["webusb"]["close"] +'</a>');
+        .catch(errorHandler);
+    
+        // WebUSB Disconnect Events
+        navigator.usb.addEventListener('disconnect', event => {
+            var error = {"message": config["translate"]["webusb"]["err"]["device-disconnected"]};
+            errorHandler(error);
         });
+
+    }
+
+    function errorHandler(err) {
+        // Disconnect
+        doDisconnect();
+
+        // Error handler
+        $("#flashing-overlay-container").css("display", "flex");
+        $("#flashing-info").addClass('hidden');
+
+
+        // If micro:bit does not support dapjs
+        var errorMessage;
+        if (err.message === "No valid interfaces found."){
+            errorMessage = "update-req";
+        } else if (err.message === "Unable to claim interface.") {
+            errorMessage = "clear-connect";
+        } else {
+            errorMessage = "reconnect-microbit";
```

Comment:
I'm a bit worried that these messages depend on DAPJs hardcoded strings in Error instance constructors (https://github.com/ARMmbed/dapjs/search?q=No+valid+interfaces+found.&unscoped_q=No+valid+interfaces+found.). So if an update in DAPJs slightly changes this text then it would break this.

Is there a different way to detect these specific errors? If not, is it possible to add some tests to check DAPJs still produces the same strings?

Reply:
We can use `err.name` instead to check the name of the DOMException

https://developer.mozilla.org/en/docs/Web/API/DOMException

I haven't got any micro:bits with an old enough version of DAPLink to test the first error, so I'm not sure which 'name' I should use. Do you have any?

```
          var errorMessage;
          if (err.message === "No valid interfaces found."){
              errorMessage = "update-req";
~         } else if (err.name === "NetworkError") {
              errorMessage = "clear-connect";
          } else {
              errorMessage = "reconnect-microbit";
          }
```

Reply:
I'll try one here, but the `No valid interfaces found.` is a normal error instance in DAPJs not derived from any specific one (`throw new Error("No valid interfaces found.")`), so the name will just be `"Error"`. It'd be nice if they had custom errors.

Reply:
Just tested it, and it is just `"Error"`:

![image](https://user-images.githubusercontent.com/29712657/63157302-8d0c9d00-c00e-11e9-9381-afaf2c29c98b.png)

Reply:
Hmm.. maybe we're gonna have to rely on the err.message rather than name, or modify DAPjs. For this PR should we stick with err.message, and create an issue to look at it properly in the next/a future sprint?

Reply:
Yeah, could you create a ticket in the platform tracker to look into DAPJs error parsing?
We'll likely submit a PR or ask DAPJs to create custom errors, but that's one for another day.